RequestService/RequestService.Handlers/PutUpdateShiftStatusToAcceptedHandler.cs

```diff
@@ -0,0 +1,112 @@
+﻿using MediatR;
+using System.Threading;
+using System.Threading.Tasks;
+using RequestService.Core.Interfaces.Repositories;
+using RequestService.Core.Services;
+using HelpMyStreet.Contracts.RequestService.Request;
+using HelpMyStreet.Contracts.RequestService.Response;
+using HelpMyStreet.Contracts.CommunicationService.Request;
+using HelpMyStreet.Utils.Enums;
+using HelpMyStreet.Utils.Models;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace RequestService.Handlers
+{
+    public class PutUpdateShiftStatusToAcceptedHandler : IRequestHandler<PutUpdateShiftStatusToAcceptedRequest, PutUpdateShiftStatusToAcceptedResponse>
+    {
+        private readonly IRepository _repository;
+        private readonly ICommunicationService _communicationService;
+        private readonly IJobService _jobService;
+        public PutUpdateShiftStatusToAcceptedHandler(IRepository repository, ICommunicationService communicationService, IJobService jobService)
+        {
+            _repository = repository;
+            _communicationService = communicationService;
+            _jobService = jobService;
+        }
+
+        public async Task<PutUpdateShiftStatusToAcceptedResponse> Handle(PutUpdateShiftStatusToAcceptedRequest request, CancellationToken cancellationToken)
+        {
+            PutUpdateShiftStatusToAcceptedResponse response = new PutUpdateShiftStatusToAcceptedResponse()
+            {
+                Outcome = UpdateJobStatusOutcome.Unauthorized,
+                JobID = -1
+            };
+
+            int existingJobId = await _repository.VolunteerAlreadyAcceptedShift(request.RequestID, request.SupportActivity.SupportActivities.First(), request.VolunteerUserID, cancellationToken);
+
+            if (existingJobId>0)
+            {
+                response = new PutUpdateShiftStatusToAcceptedResponse()
+                {
+                    Outcome = UpdateJobStatusOutcome.AlreadyInThisStatus,
+                    JobID = existingJobId
+                };
+                return response;
+            }
+            else
+            {
+                var jobId = _repository.UpdateShiftStatusToAccepted(request.RequestID, request.SupportActivity.SupportActivities.First(), request.CreatedByUserID,  request.VolunteerUserID, cancellationToken);
```

Comment:
I think we have a potential concurrency issue here - two threads (from retry logic) could both find existingJobId = -1, and hence both hit this block.  Maybe we need a database constraint that prevents a user accepting the same job twice, or a final check that a user hasn't double-accepted?

Reply:
will be tackled in https://helpmystreet.atlassian.net/jira/software/projects/VRRM/boards/25?selectedIssue=VRRM-51